keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1284 @@
+
+const (
+
+	// This is an action which might be taken on pod failure - mark the
+	// pod's job as Failed and terminate all running pods.
+	BackoffActionTerminate BackoffPolicyAction = "Terminate"
+
+	// This is an action which might be taken on pod failure - the pod will be
+	// restarted and the counter for .backoffLimit will not be incremented.
+	BackoffActionIgnore BackoffPolicyAction = "Ignore"
+
+)
+
+// BackoffPolicySource specifies the source of the failure
+type BackoffPolicySource string
+
+const (
+
+	BackoffPolicySourceExitCode BackoffPolicySource = "ExitCode"
+
+	BackoffPolicySourceStatusReason BackoffPolicySource = "StatusReason"
+
+)
+
+type BackoffPolicyRequirementOperator string
+
+const (
+	BackoffPolicyRequirementOperatorIn                      BackoffPolicyRequirementOperator = "In"
+	BackoffPolicyRequirementOperatorNotIn                   BackoffPolicyRequirementOperator = "NotIn"
+  BackoffPolicyRequirementOperatorInClosedRange           BackoffPolicyRequirementOperator = "InClosedRange"
+  BackoffPolicyRequirementOperatorNotInClosedRange        BackoffPolicyRequirementOperator = "NotInClosedRange"
+)
+
+// BackoffPolicy describes how failed pods influence the backoffLimit.
+type BackoffPolicy struct {
+	// A list of backoff policy rules. The rules are evaluated in order.
+	// Once a rule matches a Pod failure, the remaining of the rules are ignored.
+	Rules []BackoffPolicyRule
+}
+
+type BackoffPolicyRule struct {
+	// Specifies the action taken on a pod failure when the requirements are satisfied.
+	Action BackoffPolicyAction
+	// A list of backoff policy requirements. The requirements are ANDed.
+	Requirements []BackoffPolicyRequirement
+}
+
+type BackoffPolicyRequirement struct {
+	// Specifies the source of the failure. Valid sources are:
+	// - ExitCode: The exit code from a container
+	// - StatusReason: The reason for the Pod failed status.
+	Source BackoffPolicySource
+
+	// Restricts the check for exit codes to only apply to the container with
+	// the specified name. When empty the rule applies to all containers
+	// +optional
+	ContainerName *string
+
+	// Represents the source’s relationship to a set of values.
+	// Valid Operators are In, NotIn, Gt and Lt.
+	Operator BackoffPolicyRequirementOperator
+
+	// An array of string values. If the operator is In or NotIn,
+	// the values array must be non-empty. If the operator is Gt or Lt, the values
+	// array must have a single element, which will be interpreted as an integer.
+	Values []string
+}
+
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+  ...
+	// Specifies the policy of handling failed pods. In particular, it allows to
+	// specify the set of actions and conditions which need to be
+	// satisfied to take the associated action.
+	// If empty, the default behaviour applies - the counter of pod failed is
+	// incremented and it is checked against the backoffLimit
+	// +optional
+	BackoffPolicy *BackoffPolicy
```

Comment:
Surfacing discussion from offline discussion:

Should we call this FailurePolicy or PodFailurePolicy instead?